test/test_codegen.py

```diff
@@ -29,15 +29,82 @@
 import pyrometheus as pyro
 
 
+def test_kinetics():
+    """This function tests that pyrometheus-generated code
+    computes the right rates of progress for given temperature
+    and composition"""
+    sol = ct.Solution("sanDiego.cti", "gas")
+    ptk = pyro.gen_python_code(sol)()
+    # Homogeneous reactor to get test data
+    init_temperature = 1500.0
+    equiv_ratio = 1.0
+    ox_di_ratio = 0.21
+    stoich_ratio = 0.5
+    i_fu = sol.species_index("H2")
+    i_ox = sol.species_index("O2")
+    i_di = sol.species_index("N2")
+    x = np.zeros(ptk.num_species)
+    x[i_fu] = (ox_di_ratio*equiv_ratio)/(stoich_ratio+ox_di_ratio*equiv_ratio)
+    x[i_ox] = stoich_ratio*x[i_fu]/equiv_ratio
+    x[i_di] = (1.0-ox_di_ratio)*x[i_ox]/ox_di_ratio
+    sol.TPX = init_temperature, ct.one_atm, x
+    reactor = ct.IdealGasConstPressureReactor(sol)
+    sim = ct.ReactorNet([reactor])
+    time = 0.0
+    for step in range(50):
+        time += 1.0e-6
+        sim.advance(time)
+        # Cantera kinetics
+        r_ct = reactor.kinetics.net_rates_of_progress
+        omega_ct = reactor.kinetics.net_production_rates
+        # Get state from Cantera
+        temp = reactor.T
+        rho = reactor.density
+        y = reactor.Y
+        # Prometheus kinetics
+        c = ptk.get_concentrations(rho, y)
+        r_pm = ptk.get_net_rates_of_progress(temp, c)
+        omega_pm = ptk.get_net_production_rates(rho, temp, y)
+        # Print
+        err_r = np.abs((r_ct-r_pm)/r_ct).max()
+        err_omega = np.abs((omega_ct[0:-1]-omega_pm[0:-1])/omega_ct[0:-1]).max()
+        # Compare
+        assert err_r < 1.0e-10
+        assert err_omega < 1.0e-8
```

Comment:
What are the dominant parts of this error? Do we have any reason to believe that these tolerances would also work for different mechanisms?

Reply:
These tolerances were too loose and I can't recall why. One reason I suspect is I was taking relative errors of very small quantities. I've tightened them to 1e-12. This accuracy is enough to reproduce published ignition times. Also, it works for San Diego and UIUC, but I'll bring in larger mechs to test against.

Reply:
FWIW, both of these pass to 1e-10 for our current mechanisms. See e.g. same test in #7